### PR TITLE
protobuf: add protoc option

### PR DESCRIPTION
**Pull request by madebr**
Specify library name and version:  **protobuf/3.9.1.0**
- This recipe allows protobuf to be used as a build requirement.
- Add option to build protoc.

fixes #65 #1956

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by madebr** [comment]
>         * `[HOOK - conan-center.py] post_package_info(): ERROR: [CMAKE FILE NOT IN BUILD FOLDERS (KB-H019)] The *.cmake files have to be placed in a folder declared as `cpp_info.builddirs`. Currently folders declared: ['lib\\cmake\\protobuf'] (https://github.com/conan-io/conan-center-index/wiki/Error-Knowledge-Base#KB-H019)`
>         * `[HOOK - conan-center.py] post_package_info(): ERROR: [CMAKE FILE NOT IN BUILD FOLDERS (KB-H019)] Found files:`

@uilianries On Windows, the hook should normalize windows paths to cmake paths (`"\\"` -> `"/"`)

**Comment by madebr** [comment]
@jgsogo This recipe provides some libraries library (libprotobuf/libprotoc/..) and an executable (protoc).
So this recipe can be used as a build requirement and a ordinary requirement.
In `package_info` i add the `bin` folder to the `PATH` environment variable.

Won't this cause problems when cross building? Because `PATH` might be appended by a directory, containing executables for the host architecture..

**Comment by madebr** [comment]
>         * `[HOOK - conan-center.py] pre_export(): ERROR: [TEST PACKAGE - RUN ENVIRONMENT (KB-H029)] The 'RunEnvironment()' build helper is no longer needed. It has been integrated into the self.run(..., run_environment=True) (https://github.com/conan-io/conan-center-index/wiki/Error-Knowledge-Base#KB-H029) `

@uilianries  The `RunEnvironment` helper is needed in the `build()` step of the cmake script.
So either:
- the cmake helper needs to be improved (as does `AutoToolsBuildEnvironment` and `Meson`); or
- the hooks is too stringent.

**Comment by Croydon** [comment]
Maybe the hook should check if `RunEnvironment` is used inside of `test()` because there should be no reasons why it should be used there? 🤔

File: recipes/protobuf/3.9.x/conanfile.py [hunk 1]
```diff
@@ -32,57 +47,98 @@ def _is_clang_x86(self):
 
     def source(self):
         tools.get(**self.conan_data["sources"][self.version])
-        extracted_folder = self.name + "-" + self.version
-        os.rename(extracted_folder, self._source_subfolder)
+        version = re.findall("(([0-9]+\\.)*[0-9]+)", os.path.basename(os.path.basename(self.conan_data["sources"][self.version]["url"])))[0][0]
```
Comment (uilianries) [inline recipes/protobuf/3.9.x/conanfile.py]: why do you need all this complex parse to get the version?

File: recipes/protobuf/3.9.x/conanfile.py [hunk 2]
```diff
@@ -32,57 +47,98 @@ def _is_clang_x86(self):
...
+        if self.settings.os == "Windows":
+            del self.options.fPIC
 
     def configure(self):
-        if self.settings.os == "Windows" and self.settings.compiler == "Visual Studio":
+        if self.options.shared:
             del self.options.fPIC
-            compiler_version = Version(self.settings.compiler.version.value)
+        if self.settings.compiler == "Visual Studio":
+            compiler_version = tools.Version(self.settings.compiler.version.value)
```
Comment (uilianries) [inline recipes/protobuf/3.9.x/conanfile.py]:
```suggestion
            compiler_version = tools.Version(self.settings.compiler.version)
```

**Comment by jgsogo** [comment]
> @jgsogo This recipe provides some libraries (libprotobuf/libprotoc/..) and an executable (protoc).
> So this package can be used as a build requirement and a ordinary requirement.
> In `package_info` I add the `bin` folder to the `PATH` environment variable.
> 
> Won't this cause problems when cross building (and adding it in both `requirements` and `build_requirements`)? Because `PATH` might be appended by a directory, containing executables for the host architecture..

Have a look to the implementation of the feature in the sources: https://github.com/jgsogo/conan/blob/4b977e322c308341003007b5641aadc3df36383b/conans/client/installer.py#L503

 * Recipes in the host context (requirements and build_requires forced to host) propagate only the `cpp_info` to `deps_cpp_info`

   ```python
   conan_file.deps_cpp_info.update(n.conanfile.cpp_info, n.ref.name)
   ```

 * Regular build requirements (those will be in the ``build`` context) propagate the `env_info` and the `cpp_info` to the `deps_env_info`:

   ```python
   env_info = EnvInfo()
   env_info._values_ = n.conanfile.env_info._values_.copy()
   # Add cpp_info.bin_paths/lib_paths to env_info (it is needed for runtime)
   env_info.DYLD_LIBRARY_PATH.extend(n.conanfile.cpp_info.lib_paths)
   env_info.DYLD_LIBRARY_PATH.extend(n.conanfile.cpp_info.framework_paths)
   env_info.LD_LIBRARY_PATH.extend(n.conanfile.cpp_info.lib_paths)
   env_info.PATH.extend(n.conanfile.cpp_info.bin_paths)
   conan_file.deps_env_info.update(env_info, n.ref.name)
   ```

This is not the same as it was being done before. This is not breaking because using two profiles is an opt-in, but we know this can break some recipes or workflows, **we really wan't to know how people is using the propagation in the projects** and try to improve it for Conan 2.0. We might change it if needed based on your feedback.

File: recipes/protobuf/3.9.x/conandata.yml [hunk 3]
```diff
@@ -1,4 +1,8 @@
 sources:
-  3.9.1:
-    sha256: 98e615d592d237f94db8bf033fba78cd404d979b0b70351a9e5aaff725398357
-    url: https://github.com/protocolbuffers/protobuf/archive/v3.9.1.tar.gz
+  "3.9.1.0":
```
Comment (uilianries) [inline recipes/protobuf/3.9.x/conandata.yml]: You can't change the version since it's already available in CCI.

**Comment by madebr** [comment]
> Have a look to the implementation of the feature in the sources: https://github.com/jgsogo/conan/blob/4b977e322c308341003007b5641aadc3df36383b/conans/client/installer.py#L503

>       ```python
>       env_info = EnvInfo()
>       env_info._values_ = n.conanfile.env_info._values_.copy()
>       # Add cpp_info.bin_paths/lib_paths to env_info (it is needed for runtime)
>       env_info.DYLD_LIBRARY_PATH.extend(n.conanfile.cpp_info.lib_paths)
>       env_info.DYLD_LIBRARY_PATH.extend(n.conanfile.cpp_info.framework_paths)
>       env_info.LD_LIBRARY_PATH.extend(n.conanfile.cpp_info.lib_paths)
>       env_info.PATH.extend(n.conanfile.cpp_info.bin_paths)
>       conan_file.deps_env_info.update(env_info, n.ref.name)
>       ```

@jgsogo 
Thanks for the reference code.
Adding these extra fields to `cpp_info`seems like a good idea.
Can a recipe detect in its `package_info` in what context it is used? (build/host context)

**Comment by uilianries** [comment]
> 
> The `RunEnvironment` helper is needed in the `build()` step of the cmake script.
> So either:
> 
>     * the cmake helper needs to be improved (as does `AutoToolsBuildEnvironment` and `Meson`); or
> 
>     * the hook is too stringent.
> 
> 
> This is the same problem as @ericLemanissier was seeing while packaging libsndfile.
> See [#1103 (comment)](https://github.com/conan-io/conan-center-index/pull/1103#discussion_r394042553)

I'll push a new fix for hooks, thanks for reporting.

File: recipes/protobuf/3.9.x/conanfile.py (hunk 1 above)
Comment (madebr) [reply recipes/protobuf/3.9.x/conanfile.py]: The version of the tar.gz is 3.9.1 but the protoc binary has a 3.9.1.0 suffix.
But I'll change the version back to 3.9.1 and append the .0 to the binary.

File: recipes/protobuf/3.9.x/conandata.yml (hunk 3 above)
Comment (madebr) [reply recipes/protobuf/3.9.x/conandata.yml]: I'll revert it.

Comment (uilianries) [reply recipes/protobuf/3.9.x/conandata.yml]: thanks!

**Comment by jgsogo** [comment]
@madebr 

> Adding these extra fields to `cpp_info` seems like a good idea.

We are not adding extra to `cpp_info`, with the two-profiles approach we are propagating (for the build requires in the build context) the `cpp_info` and the `env_info` to the `deps_env_info`, and before entering the `build()` method of the consumers all those variables will populate the environment: `DYLD_...`, `LD_...`, `PATH_...`.

But, the `deps_cpp_info` in the consumers will contain less information.

> Can a recipe detect in its `package_info` in what context it is used? (build/host context)

Right now (v1.24) it can't. This is something we need to investigate and implement for the next releases.

File: recipes/protobuf/3.9.x/conanfile.py (hunk 1 above)
Comment (madebr) [reply recipes/protobuf/3.9.x/conanfile.py]: @uilianries Fixed it. Please review

**Comment by madebr** [comment]
@uilianries Please re-review.
I've added protoc support to protobuf/3.11.4 as well.

**Comment by madebr** [comment]
@danimtb @jgsogo @uilianries @SSE4 
Someone with a mac, please help me  :pray: :smile:

**Comment by jgsogo** [comment]
It works in static, it is failing in shared because it is not able to find a shared library. The problem is similar to this one here https://github.com/conan-io/examples/pull/17#pullrequestreview-401928903

This should be the `protoc` running in the `protobuf_generate_cpp` call it is not able to find the `dylib`.  We need to populate the `DYLD...` variable in order to call it. The information is in the Conan generated files, I'll try to find how to _inject_ that variable into the environment.

**Comment by jgsogo** [comment]
Ok, this line in `protobuf-generate.cmake` works:

```cmake
    add_custom_command(
      OUTPUT ${_generated_srcs}
      COMMAND ${CMAKE_COMMAND} -E env "DYLD_LIBRARY_PATH=${CONAN_LIB_DIRS}" protoc
      ARGS --${protobuf_generate_LANGUAGE}_out ${_dll_export_decl}${protobuf_generate_PROTOC_OUT_DIR} ${_protobuf_include_path} ${_abs_file}
      DEPENDS ${_abs_file}
      COMMENT "Running ${protobuf_generate_LANGUAGE} protocol buffer compiler on ${_proto}"
      VERBATIM )
```

The problem here: depending on the generator, the variable `${CONAN_LIB_DIRS}` might not be available (it comes from the `cmake` generator, it is `Protobuf_LIB_DIRS` for `cmake_find_package`). And getting it from the target is also a bit painful because it doesn't have the same name.

**Comment by jgsogo** [comment]
We can concat paths coming from different generators (beware of the path separator for different OS): 

```
COMMAND ${CMAKE_COMMAND} -E env "DYLD_LIBRARY_PATH=${Protobuf_LIB_DIRS}:${CONAN_LIB_DIRS}" protoc
```

More complex logic if we want to cover `_multi` generators too.

**Comment by madebr** [comment]
> ```cmake
>       COMMAND ${CMAKE_COMMAND} -E env "DYLD_LIBRARY_PATH=${CONAN_LIB_DIRS}" protoc
> ```
> The problem here: depending on the generator, the variable `${CONAN_LIB_DIRS}` might not be available (it comes from the `cmake` generator, it is `Protobuf_LIB_DIRS` for `cmake_find_package`). And getting it from the target is also a bit painful because it doesn't have the same name.

Thanks @jgsogo !
In the `test_package`, the `DYLD_LIBRARY_PATH` variable was set using the `RunEnvironment` helper.
But I didn't see that the helper overrides this variable!

I'll (probably) only need to patch out setting that env var.

Great find!

**Comment by madebr** [comment]
@jgsogo 
I'm hesitant about adding `"${CMAKE_COMMAND}" -E env "DYLD_LIBRARY_PATH=XXXX" protoc ARG=...` to the cmake script.

I think the problem on mac is more fundamental.
In `test_package/conanfile.py`, the `RunEnvironment` helper is used.
So the `DYLD_LIBRARY_PATH` is set.

But in the cmake script, this variable is empty.
On Linux, this works fine.

I'm not sure `tools.environment_append` is working correctly on Macos.

**Comment by jgsogo** [comment]
The problem is about the "System Integrity Protection" in MacOS, you need to set the environment "DYLD_LIBRARY_PATH" in the call to the subprocess, so you need CMake to add that variable to the environment running the command.

This is why in Conan `self.run(".....", run_environment=True)` works, while `with self.environmnt_append(): self.run("...")` doesn't work. In the first call we are creating thee command line command like: `DYLD_LIBRARY_PATH=xxxxx <command>"

**Comment by madebr** [comment]
Thanks for giving more background.

So the the build helpers (CMake, AutoToolsEnvironment, ...) helpers need a `run_environment` argument, such that `CMake._run` can pass it to `ConanFile.run`.

Would this solve this problem?

**Comment by jgsogo** [comment]
For this `add_custom_command` we need to modify the `.cmake` file, it is the same when running tests using `cmake --build ... --target test` (https://github.com/conan-io/examples/pull/17#pullrequestreview-401928903). From outside CMake (where Conan lives) we cannot add anything to the environment, it won't be used in the subprocess.

**Comment by madebr** [comment]
> For this `add_custom_command` we need to modify the `.cmake` file, it is the same when running tests using `cmake --build ... --target test` ([conan-io/examples#17 (review)](https://github.com/conan-io/examples/pull/17#pullrequestreview-401928903)). From outside CMake (where Conan lives) we cannot add anything to the environment, it won't be used in the subprocess.

Ok, I'll patch the cmake script.

The original code used the cmake imported target `protobuf::protoc`.
If the new conan components feature would be able to generate this executable target, it can also be used.
Would this be another way to solve this?

**Comment by madebr** [comment]
For the `cmake_find_package_multi` generator, I think we need to have the [INTERFACE_LINK_DIRECTORIES](https://cmake.org/cmake/help/latest/prop_tgt/INTERFACE_LINK_DIRECTORIES.html) property of the libraries populated.
Then we can use generator expressions to extract the property of the correct build type.

Now I would need to add `${Protobuf_LIB_DIRS_RELEASE}:${Protobuf_LIB_DIRS_DEBUG}:${Protobuf_LIB_DIRS_RELWITHDEBINFO}:${Protobuf_LIB_DIRS_MINSIZEREL}`.
And then, this doesn't guarantee success if a user added an extra build type to its `settings.yml`.

**Comment by Croydon** [comment]
I don't understand if there are any issues left, but if they aren't breaking and can't be fixed quickly I would suggest to create an issue to fix it later and finally merge this,

This becomes quite a blocker IMO

**Comment by uilianries** [comment]
@Croydon Agreed, Protobuf is one of most required packages

**Comment by madebr** [comment]
> Agree, we can move this one forward and wait for issues.

I love your optimism :smile:

**Comment by jgsogo** [comment]
> > Agree, we can move this one forward and wait for issues.
> 
> I love your optimism 😄

🤣  well, I didn't mean there will be issues.... but, well, we all know there will be. I'll spend some days now experiencing with the cross-building feature, env propagation, settings,... so I'll have time to try this recipe and open the issues myself 😄 

homer-seto

**Comment by Croydon** [comment]
Thanks everyone! 🥳

**Pull request merged**